### ISSUE TITLE
dsnoop/dmix capture/playout hangs on daemon restart

### ISSUE THREAD

**Issue by AleksanderZdunek** [bondagit/aes67-linux-daemon#82]
When receiving audio on a dsnoop pcm the capture hangs indefinitely when the aes67 daemon is restarted. When using hw:RAVENNA or plughw:RAVENNA pcm the capture times out after daemon restart.
Same problem occurs when playing out audio on a dmix pcm. After aes67 daemon restarts a dmix playout hangs indefinitely and a hw playout times out.

A minimum reproducible example can be seen using arecord, speaker-test and the default dsnoop/dmix pcms.

`arecord -D hw:RAVENNA -r 48000 -f S32_LE -c 2 -t raw | hexdump`
After daemon restart arecord times out after 10 seconds with output:
`arecord: pcm_read:2152: read error: Input/output error`

`arecord -D dsnoop:RAVENNA -r 48000 -f S32_LE -c 2 -t raw | hexdump`
When using dsnoop arecord hangs indefinitely after daemon restart.

`speaker-test -D hw:RAVENNA -r 48000 -F S32_LE -c 2 -t sine`
After daemon restart speaker-test times out after 10 seconds with output:
```
Write error: -5,Input/output error
xrun_recovery failed: -5,Input/output error
Transfer failed: Input/output error
```

`speaker-test -D dmix:RAVENNA -r 48000 -F S32_LE -c 2 -t sine`
When using dmix speaker-test hangs indefinitely after daemon restart.

I have an application that routes audio between various audio interfaces. I'm trying to integrate AES67 virtual audio interfaces.
I'm using custom dsnoop and dmix definitons, but the problem seem to be the same with the default pcms.
Needing to restart the routing application after the aes67 daemon has started or restarted is suboptimal.

Is there a way to get dsnoop/dmix to resume capture/playout after aes67 daemon restart? Alternatively, make them time out like the hw pcm, so that the routing application can recover?

**Comment by bondagit** [comment]
When the daemon is terminated gracefully it closes all driver streams and this explains why you see this behaviour.
Why do you need to restart the daemon ?

**Comment by AleksanderZdunek** [comment]
Changing settings in the web interface restarts the daemon. It can also happen accidentally by user error or conceivably by a system error.
I'm also currently having a problem with startup order after a machine reboot.
In all such cases I would like the solution to be more robust.

**Comment by bondagit** [comment]
> Changing settings in the web interface restarts the daemon. 

Yes, at the moment changing any of the main config settings restarts the daemon and this in turn stops the audio streams.
Improvements in this area are certainly possible because not all the parameters would require a full restart: for example changing any of the SAP or Logging parameters cloud be handled without a full restart.
Anyway for most of the parameters full streams re-configuration is needed because these parameters are actually passed to the kernel module: for example a change in the RTP base address or port and especially in the TIC frame size requires streams re-programming that cannot be done on the fly and that would require in any case a stream stop.

> It can also happen accidentally by user error
We could improve the WebUI to show up a modal dialog, when a config parameter change is applied, to inform the user about the daemon restart and ask for confirmation to proceed.

> conceivably by a system error.
I don't see as this can happen. I am not aware of any bug causing an unexpected daemon termination.

> I'm also currently having a problem with startup order after a machine reboot.
Can you please elaborate. What kind of problem do you have ? 
At startup time you have to install the module first and start the daemon at any time just after that.
On termination you have to terminate the daemon first and remove the driver afterwards.

**Comment by AleksanderZdunek** [comment]
It seems as a minimum workaround I need to somehow detect that the kernel module has stopped the stream and restart the routing application to reinitialize all streams.

As mentioned, I have tested using hw:RAVENNA pcm which apparently lets the routing application detect by way of read/write timeout that a stream has stopped. It can then recover by restarting the capture/playout thread for the stream. This does not work when using dsnoop/dmix pcms. Not being able to use dsnoop/dmix complicates support for 64 channels for me. dsnoop/dmix fits neatly into existing architecture. 

> We could improve the WebUI to show up a modal dialog
Like you say, this wouldn't solve the problem for legitimate cases of daemon restart. Reconfiguration happens very frequently during initial setup and learning phase, which I can personally attest to :)
The long term plan is anyway to build a new interface that would integrate more seamlessly into existing UI.
As a side note though, I have been quite impressed with your WebUI. It has been nice to work with during testing.

>Can you please elaborate. What kind of problem do you have ?
After setting up an audio flow over a three node test network the flow does not survive rebooting a node. After reboot I need to then manually restart the daemon and the routing application, in that order, for audio to resume. Presumably this can be solved by ensuring that the module is loaded and daemon running before starting the routing application on boot, but I haven't had time to look into this end of the problem yet. The goal here is to make the node automatically resume configured audio routing duties on startup, to make it robust for users.

**Comment by bondagit** [comment]
I have elaborated a bit other the issues you reported and it's possible to apply the following improvements:
- on config setting change, if possible, the daemon will perform a light restart not stopping the streams. As already mentioned this is not always possible and on some specific configuration change (as for example the TIC frame size) full streams restart is required.
- in order to better support the activation of the playback and capture applications the daemon will invoke a shell script when the PTP gets locked and unlocked. In general the playback and capture is only possible when the PTP is locked because in that case the audio clock is running. When the PTP is not locked the playback and capture hang and this, depending if the applications are using the RAVENNA device directly or a plugin can cause an I/O error or an hang (as you reported).

**Comment by bondagit** [comment]
Automatic load and unload of the kernel module can be usually archived using modprobe and related configuration files: /etc/modules-load.d/* . See `man 5 modules-load.d`
Automatic startup and stop of the daemon is usually achievable by using _Init_ or _Systemd_ configuration files depending on your distro.
These configurations are out of scope here as they depend on your distro.

**Comment by AleksanderZdunek** [comment]
> on config setting change, if possible, the daemon will perform a light restart not stopping the streams. As already mentioned this is not always possible and on some specific configuration change (as for example the TIC frame size) full streams restart is required.

This sounds like a nice thing to have.

> in order to better support the activation of the playback and capture applications the daemon will invoke a shell script when the PTP gets locked and unlocked. In general the playback and capture is only possible when the PTP is locked because in that case the audio clock is running. When the PTP is not locked the playback and capture hang and this, depending if the applications are using the RAVENNA device directly or via a plugin, can cause an I/O error or an application hang (as you reported). The shell script will offer the opportunity to start or stop the player/s according to the PTP status.

I came up with something along that line, although without the script hook, as a workaround. Tying the hook to PTP lock is good to know.

Thanks for the feedback.

**Comment by bondagit** [comment]
Please checkout [issue_82 branch](https://github.com/bondagit/aes67-linux-daemon/tree/issue_82) and give it a try.
Info are provided by the commit log.

**Comment by AleksanderZdunek** [comment]
Seems to be working well. Thanks!

**Comment by bondagit** [comment]
> Seems to be working well. Thanks!

Good. Thanks for the feedback.
Nice thing is that a Dante receiver seems to be able to accept stream reconfiguration dynamically: so, if for example I change the RTP multicast base address and/or port and I apply the settings on the daemon, the SAP session gets updated, the player doesn't stop, and Dante is able to continue the reception.

I will continue testing for a while and if everything is OK I will merge he changes with master branch.

**Comment by bondagit** [comment]
Improvements have been delivered with daemon v1.5
